observability: Add multicluster-observability-addon proposal (#1613)
openshift/enhancements: enhancements/observability/multicluster-observability-addon.md

        'tls.key': "Base64 endoded TLS key"
        'ca-bundle.crt': "Base64 encoded Certificate Authority certificate"
    - apiVersion: logging.openshift.io/v1
      kind: ClusterLogForwarder
      metadata:
        name: mcoa-instance
        namespace: openshift-logging
      spec:
        outputs:
        - loki:
            labelKeys:
            - log_type
            - kubernetes.namespace_name
            - kubernetes.pod_name
            - openshift.cluster_id
          name: application-logs
          type: loki
          url: "https://loki-foo.grafana.com"
          secret:
            name: application-logs-loki-mtls-credentials
        - name: cluster-logs
          type: cloudwatch
          cloudwatch:
            groupBy: logType
            region: eu-central-1
          secret:
            name: cluster-logs-aws-cloudwatch-credentials
        pipelines:
        - inputRefs:
          - application
          name: application-logs
          outputRefs:
          - application-logs
        - inputRefs:
          - audit
          - infrastructure
          name: cluster-logs
          outputRefs:
          - cluster-logs
```

> #### Multi Cluster OTLP Collection and Forwarding

Author: @iblancasa could you please add a commit to this PR updating the resources for the tracing side? I'm mainly concerned with an up-to-date example of the `OpenTelemetryCollector` and the resulting resource in the `ManifestWorks`

---
title: multicluster-logs-traces-forwarding
authors:
  - "@periklis"
  - "@JoaoBraveCoding"
  - "@pavolloffay"
  - "@iblancasa"
reviewers:
  - "@alanconway"
  - "@jcantrill"
  - "@xperimental"
  - "@jparker-rh"
  - "@berenss"
  - "@bjoydeep"
approvers:
  - "@alanconway"
  - "@jcantrill"
  - "@xperimental"
api-approvers:
  - "@alanconway"
  - "@jcantrill"
  - "@xperimental"
creation-date: 2023-11-28
last-updated: 2023-11-28
tracking-link:
  - https://issues.redhat.com/browse/OBSDA-356
  - https://issues.redhat.com/browse/OBSDA-393
  - https://issues.redhat.com/browse/LOG-4539
  - https://issues.redhat.com/browse/OBSDA-489
see-also:
  - None
replaces:
  - None
superseded-by:
  - None
---

# Multi-Cluster Logs and Traces Forwarding

## Release Signoff Checklist

> - [x] Enhancement is `implementable`
> - [ ] Design details are appropriately documented from clear requirements
> - [ ] Test plan is defined
> - [ ] Graduation criteria for dev preview, tech preview, GA

Reviewer: ```suggestion
- [x] Enhancement is `implementable`
- [x] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
```

        name: cluster-logging
        source: redhat-operators
        sourceNamespace: openshift-marketplace
        startingCSV: cluster-logging.v5.9.0
    - apiVersion: v1
      kind: Secret
      metadata:
        name: aws-credentials
        namespace: openshift-logging
      data:
        access_key_id: "Base64 access key id"
        access_key_secret: "Base64 access key secret"
    - apiVersion: v1
      kind: Secret
      metadata:
        name: loki-foo-mTLS-credentials
        namespace: openshift-logging
      data:
        'tls.crt': "Base64 encoded TLS client certificate"
        'tls.key': "Base64 endoded TLS key"
        'ca-bundle.crt': "Base64 encoded Certificate Authority certificate"
    - apiVersion: logging.openshift.io/v1
      kind: ClusterLogForwarder
      metadata:
        name: instance
        namespace: open-cluster-management
      spec:
        outputs:
        - loki:
            labelKeys:
            - log_type
            - kubernetes.namespace_name
            - kubernetes.pod_name
            - openshift.cluster_id
          name: application-logs
          type: loki
          url: "https://loki-foo.grafana.com"
          secret:

>             name: loki-foo-mTLS-credentials

Reviewer: ```suggestion
            name: application-logs-loki-mtls-credentials
```

        - name: cluster-logs
          type: cloudwatch
          cloudwatch:
            groupBy: logType
            region: eu-central-1
          secret:

>             name: aws-credentials

Reviewer: ```suggestion
            name: cluster-logs-aws-cloudwatch-credentials
```

        pipelines:
        - inputRefs:
          - application
          name: application-logs
          outputRefs:
          - application-logs
        - inputRefs:
          - audit
          - infrastructure
          name: cluster-logs
          outputRefs:
          - cluster-logs
```

#### Multi Cluster OTLP Collection and Forwarding

> For all managed clusters the fleet administrator is required to provide, at least, a single `OpenTelemetryCollector` resource stanza that describes the OpenTelemetry forwarding configuration for the entire fleet in the default namespace `open-cluster-management`.

Reviewer: ```suggestion
For all managed clusters the fleet administrator is required to provide, at least, a single `OpenTelemetryCollector` resource stanza that describes the OpenTelemetry forwarding configuration for the entire fleet in the default namespace `open-cluster-management-observability`.
```

The following example resource describes a configuration where a `OpenTelemetryCollector` instance is deployed per spoke cluster and it sends data to a Hub OTEL Cluster (note that this cluster can be different from the RHACM Hub cluster). The Hub OTEL Cluster then exports the received telemetry to a third-party storage solution.

```mermaid
flowchart TD
    subgraph Spoke cluster
    A[OTEL Collector] 
    end
    subgraph Spoke cluster
    B[OTEL Collector] 
    end
    subgraph Spoke cluster
    C[OTEL Collector] 
    end
    subgraph Hub OTEL Cluster
    D[OTEL Collector] 
    E[Grafana Tempo]
    end
    A --> |otlp|D
    B --> |otlp|D
    C --> |otlp|D
    D --> E
    subgraph Third party services
    D --> F[Dynatrace]
    D --> G[Datadog]
    end
```

The following example resource describes a configuration for forwarding application otlp data from one OpenTelemetry Collector (deployed in the spoke cluster) to the OpenTelemetry Hub instance:

```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: spoke-otelcol

>   namespace: open-cluster-management

Reviewer: ```suggestion
  namespace: open-cluster-management-observability
```

spec:
  config: |
    receivers:
      jaeger:
        protocols:
          grpc:
      otlp:
        protocols:
          grpc:
          http:
    processors:
    exporters:
      debug:
      otlp:
        tls:
          insecure: true
    service:
      pipelines:
        traces:
          receivers: [jaeger,otlp]
          processors: []
          exporters: [otlp, debug]
```

For the `otlp` exporter MCOA will provide a Secret `tracing-otlp-auth` with all required authentication information, e.g. TLS client Certificate:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: tracing-otlp-auth
  namespace: spoke
stringData:
  'tls.crt': "TLS client certificate"
  'tls.key': "TLS key"
  'ca-bundle.crt': "Certificate Authority certificate" # Pull from the referenced Secret otel-gateway
```

In turn the addon will compile a `ManifestWork` for the managed cluster `spoke` as follows and pass it over it's WorkAgentController:

```yaml
kind: ManifestWork
metadata:
  name: addon-multicluster-observability-addon-deploy-0
  namespace: spoke
spec:
  workload:
    manifests:

>     - apiVersion: project.openshift.io/v1
>       kind: Project
>       metadata:
>         name: openshift-opentelemetry-operator

Reviewer: Aligning this with the current charts in [here](https://github.com/rhobs/multicluster-observability-addon/tree/main/internal/addon/manifests/charts/mcoa/charts/tracing/templates) (cc @iblancasa):

```suggestion
    - apiVersion: v1
      kind: Namespace
      metadata:
        name: openshift-opentelemetry-operator
    - apiVersion: v1
      kind: Namespace
      metadata:
        name: mcoa-opentelemetry
```

---
title: multicluster-observability-addon
authors:
  - "@periklis"
  - "@JoaoBraveCoding"
  - "@pavolloffay"
  - "@iblancasa"
reviewers:
  - "@alanconway"
  - "@jcantrill"
  - "@xperimental"
  - "@jparker-rh"
  - "@berenss"
  - "@bjoydeep"
approvers:
  - "@alanconway"
  - "@jcantrill"
  - "@xperimental"
api-approvers:
  - "@alanconway"
  - "@jcantrill"
  - "@xperimental"
creation-date: 2023-11-28
last-updated: 2023-11-28
tracking-link:
  - https://issues.redhat.com/browse/OBSDA-356
  - https://issues.redhat.com/browse/OBSDA-393
  - https://issues.redhat.com/browse/LOG-4539
  - https://issues.redhat.com/browse/OBSDA-489
see-also:
  - None
replaces:
  - None
superseded-by:
  - None
---

> # Multi-Cluster Observability Addon

Reviewer: As per now we intend and plan to deliver the whole addon (collection and storage) in separate stages, we should rename this proposal to what it actually is: `Multi-Cluster Logs and Traces Forwarding`.

Reviewer: I would say no for now because metric collection is part of MCO Operator. @moadz and et.al. need to outline the migration path and in turn which CRDs to use to metric collection in MCOA.

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

> Multi-Cluster Observability has been an integrated concept in Red Hat Advanced Cluster Management (RHACM) since its inception but only incorporates one of the core signals, namely metrics, to manage fleets of OpenShift Container Platform (OCP) based clusters (See [RHACM Multi-Cluster-Observability-Operator (MCO)](rhacm-multi-cluster-observability)). The underlying architecture of RHACM observability consists of a set of observability components to collect a dedicated set of OCP metrics, visualizing them and alerting on fleet-relevant events. It is an optional but closed circuit system applied to RHACM managed fleets without any points of extensibility.

Reviewer: ```suggestion
Multi-Cluster Observability has been an integrated concept in Red Hat Advanced Cluster Management (RHACM) since its inception but only incorporates one of the core signals, namely metrics, to manage fleets of OpenShift Container Platform (OCP) based clusters (See [RHACM Multi-Cluster-Observability-Operator (MCO)](rhacm-multi-cluster-observability)). The underlying architecture of RHACM observability consists of a set of observability components to collect a dedicated set of OCP metrics, visualizing them and alerting on fleet-relevant events.
```

> This enhancement proposal seeks to bring a unified approach to collect and forward logs and traces from a fleet of OCP clusters based on the RHACM addon facility (See Open Cluster Management (OCM) [addon framework](ocm-addon-framework)) by enabling these signals events to land on third-party managed and centralized storage solutions (e.g. AWS Cloudwatch, Google Cloud Logging). The multicluster-observability-addon is an optional RHACM addon. It is a day two companion for MCO and does not necessarily share any resources/configuration with the latter. It provides a unified installation approach of required dependencies (e.g. operator subscriptions) and resources (custom resources, certificates, CA Bundles, configuration) on the managed clusters to collect and forward logs and traces. The addon's name is MultiCluster Observability Addon (MCOA).

Reviewer: ```suggestion
This enhancement proposal seeks to bring a unified approach to collect and forward logs and traces from a fleet of OCP clusters based on the RHACM addon facility (See Open Cluster Management (OCM) [addon framework](ocm-addon-framework)) by enabling these signals events to land on third-party managed storage solutions (e.g. AWS Cloudwatch, Google Cloud Logging, Splunk). The proposed addon is an optional RHACM component. It is productized and installed along with the MCO Operator. However it does neither share any resources, configuration nor depend on it. It provides a unified installation approach of required dependencies (e.g. operator subscriptions) and resources (custom resources, certificates, CA Bundles, configuration) on the managed clusters to collect and forward logs and traces. The addon's name is Multi-Cluster-Observability-Addon (short MCOA).
```

> This enhancement proposal seeks to bring a unified approach to collect and forward logs and traces from a fleet of OCP clusters based on the RHACM addon facility (See Open Cluster Management (OCM) [addon framework](ocm-addon-framework)) by enabling these signals events to land on third-party managed and centralized storage solutions (e.g. AWS Cloudwatch, Google Cloud Logging). The multicluster-observability-addon is an optional RHACM addon. It is a day two companion for MCO and does not necessarily share any resources/configuration with the latter. It provides a unified installation approach of required dependencies (e.g. operator subscriptions) and resources (custom resources, certificates, CA Bundles, configuration) on the managed clusters to collect and forward logs and traces. The addon's name is MultiCluster Observability Addon (MCOA).

Reviewer 2: Cross-linking https://github.com/rhobs/multicluster-observability-addon/pull/39#issuecomment-2065235658

There are ongoing discussions to rename tracing to opentelemetry. The objective is to provide opentelemetry capabilities via the addon.

Reviewer: @JoaoBraveCoding simply waits until this resolves, you can contribute your formulation here if you want.

## Motivation

> The main driver for the following work is to simplify and unify the installation of log and trace collection and forwarding on an RHACM managed fleet of OCP clusters. The core utility function of the addon is to install required operators (i.e. [Red Hat OpenShift Logging](ocp-cluster-logging-operator) and [Red Hat OpenShift distributed tracing platform](opentelemetry-operator)), configure required custom
> resources (i.e. `Clusterlogging`, `ClusterLogForwarder`, `OpenTelemetryCollector`) and reconcile per-cluster-related companion resources (i.e. `Secrets`, `ConfigMaps` for per-cluster authentication and configuration). This enables centralized control of the fleet's logs and traces collection and forwarding capabilities.

Reviewer: ```suggestion
The main goal here is to simplify and unify the installation of logs and traces collection and forwarding on an RHACM managed fleet of OCP clusters. The core function of the addon is to install required operators (i.e. [Red Hat OpenShift Logging](ocp-cluster-logging-operator) and [Red Hat OpenShift distributed tracing platform](opentelemetry-operator)), configure required custom
resources (i.e. `ClusterLogForwarder`, `OpenTelemetryCollector`) and reconcile per-cluster-related authentication resources (i.e. `Secrets`, `ConfigMaps`). This enables fleet-wide control of logs and traces collection and forwarding.
```

* As a fleet administrator, I want to centrally control authentication credentials for the target log forwarding outputs.
* As a fleet administrator, I want to centrally control authentication credentials for the target trace exporters.

### Goals

* Provide an optional RHACM addon to control collection and forwarding of logs and traces on managed OCP clusters.
* Enable control of authentication and authorization of storage endpoints from the hub cluster.
* Enable TLS Certificate management local on the hub cluster for managed cluster TLS client certificates.

### Non-Goals

* Provide end-to-end experience to collect, forward, store and visualize logs and traces on the same RHACM fleet.
* Share MCO's ingestion and querying capabilities for logs and traces collected by MCOA.

## Proposal

The following sections describe in detail the required resources as well as the workflow to enable collection and forwarding of logs and traces on an RHACM managed fleet of OCP clusters.

### Workflow Description

The workflow implemented in this proposal enables fleet-wide collection and forwarding of logs and traces as follows:

> 1. MCOA is installed by MCO.

Reviewer: ```suggestion
1. When the fleet administrator creates a `MultiClusterHub` resource, then MCOA is installed together with the MCO Operator.
```

   - group: logging.openshift.io
     resource: clusterlogforwarders
     # The default config is the main stanza of a ClusterLogForwarder resource
     # that describes where logs should be forwarded for all managed cluster.
     defaultConfig:
       name: instance
       namespace: open-cluster-management

   # Describes the default OpenTelemetryCollector type applied to all managed clusters.
   - group: opentelemetry.io
     resource: opentelemetrycollectors
     # The default config is the main stanza of an OpenTelemetryCollector resource
     # that describes where traces should be forwarded for all managed cluster.
     defaultConfig:
       name: spoke-otelcol
       namespace: open-cluster-management
```

and a default `AddonDeploymentConfig` describing the operator subscriptions to be used on each managed cluster:

```yaml
apiVersion: addon.open-cluster-management.io/v1alpha1
kind: AddOnDeploymentConfig
metadata:
  name: multicluster-observability-addon
  namespace: open-cluster-management
spec:
  customizedVariables:
    - name: clusterLoggingChannel
      value: stable-5.9
    - name: otelChannel
      value: stable
```

> #### Multi Cluster Log Collection and Forwarding

Reviewer: Can we add a mermaid flowchart here to visualize the input resources on the hub to the output resources on the spokes?

        installPlanApproval: Automatic
        name: opentelemetry-product
        source: redhat-operators
        sourceNamespace: openshift-marketplace
        startingCSV: opentelemetry-operator.v0.81.1-5
    - apiVersion: v1
      kind: Secret
      metadata:
        name: tracing-otlp-auth
        namespace: spoke
      stringData:
        'tls.crt': "TLS client certificate"
        'tls.key': "TLS key"
        'ca-bundle.crt': "Certificate Authority certificate"
    - apiVersion: opentelemetry.io/v1alpha1
      kind: OpenTelemetryCollector
      metadata:
        name: spoke-otelcol
        namespace: open-cluster-management
      spec:
        config: |
          receivers:
            jaeger:
              protocols:
                grpc:
            otlp:
              protocols:
                grpc:
                http:
          processors:
          exporters:
            debug:
            otlp:
              tls:
                insecure: true
          service:
            pipelines:
              traces:
                receivers: [jaeger,otlp]
                processors: []
                exporters: [otlp, debug]
```

#### Hypershift [optional]

> TBD

Reviewer: ```suggestion
N/A
```

### Drawbacks

> TBD

Reviewer: Can we describe very short what's the drawback of not having a new custom resource or extending the MultiClusterObservability resource? Also why we are not doing this at this stage?

## Design Details

### Open Questions [optional]

The [addon-framework](ocm-addon-framework) is still lacking some features that will fully enable the behaviours described in this proposal namelly: 
- [ ] Support for reconcile resources generated by the addon (Jira issues [LOG-5343](https://issues.redhat.com/browse/LOG-5343) and [TRACING-4148](https://issues.redhat.com/browse/TRACING-4148)). MCOA needs this otherwise it will not react on changes to those resources.
- [ ] Support for multiple resources of the same kind in the ManagedClusterAddOn resource (Jira issues [LOG-5344](https://issues.redhat.com/browse/LOG-5344)). MCOA needs this otherwise it will not react on changes to those resources.

### Test Plan

TBD

### Graduation Criteria

TBD

#### Dev Preview -> Tech Preview

TBD

#### Tech Preview -> GA

> TBD

Reviewer: It's time to outline the graduation criteria here, short listed as:

1. Dev Preview: CLF/OTelCollector CRs, mTLS auth & static auth, Per-ClusterSet Forwarding.
2. Tech-Preview: Refined CLF/OTelCollector CRs usage, managed auth types, HealthChecking per Signal
3. GA: TBD

#### Removing a deprecated feature

None

### Upgrade / Downgrade Strategy

None

### Version Skew Strategy

None

### Operational Aspects of API Extensions

TBD

#### Failure Modes

> TBD

Reviewer: Is it possible to describe the different failure modes on applying the wrong inputs and how this affects ManifestWorks generation? Let's not go deep, just simply listing how a missing annotation or resource affects the whole addon reconciliation loop. Maybe adding an empty section to dedicate it to health checking.

#### Support Procedures

TBD

## Implementation History

> TBD

Reviewer: I suggest to add a table here to make each stage visual, e.g. we have already a demo stage implemented.

## Alternatives

> TBD

Reviewer: I suggest to copy and adapt the alternatives as described in the EPIC description: https://issues.redhat.com/browse/LOG-4539

TBD

#### Failure Modes

TBD

#### Support Procedures

TBD

## Implementation History

TBD

## Alternatives

TBD

## Infrastructure Needed [optional]

None

[ocm-addon-framework]:https://github.com/open-cluster-management-io/addon-framework
[ocp-cluster-logging-operator]:https://github.com/openshift/cluster-logging-operator
[ocp-clusterlogforwarder-outputsecretspec]:https://github.com/openshift/cluster-logging-operator/blob/627b0c7f8c993f89250756d9601d1a632b024c94/apis/logging/v1/cluster_log_forwarder_types.go#L226-L265
[ocp-clusterlogforward-outputtypespec]:https://github.com/openshift/cluster-logging-operator/blob/627b0c7f8c993f89250756d9601d1a632b024c94/apis/logging/v1/output_types.go#L21-L40
[opentelemetry-collector-auth]:https://opentelemetry.io/docs/collector/configuration/#authentication
[opentelemetry-operator]:https://github.com/open-telemetry/opentelemetry-operator
[rhacm-multi-cluster-observability]:https://github.com/stolostron/multicluster-observability-operator

## RANDOM  IDEAS 
- What happens if we need to reference multiple static secrets?
- Consolidate annotation/labels, namely authentication annotations can/should be uniform

> - Maybe we can use annotations for the authentication methods instead of having to create a seperate resouce for that

Reviewer: > @periklis Here are some of the improvements I want to make to the EP, let me know your thoughts
> 
> > * What happens if we need to reference multiple static secrets?
> 
> Currently, if an output uses `StaticAuthentication` we look for a secret called `static-authentication` under `open-cluster-management` however this is not a correct assumption. A user might want to use different auth for a different clusterset. My idea would be that the users has to create the secret and add a label/annotation that will identify the secret. Then this secret will go on the list of resources in `ManagedClusterConfig`
> 

Yes that's probably a hindsight of my initial proposal writing that did not consider implications of per-clusterset configuration.

> > * Maybe we can use annotations for the authentication methods instead of having to create a separate resource for that
> 
> Currently, we require users to create a configmap with `output: authentication` however I think we can move this data to the annotations of the stanza. This way it's one less resource the user has to create and it help centralize the configuration. WDYT?
> 

SGTM.

> > * Consolidate annotation/labels, namely authentication annotations can/should be uniform
> 
> [rhobs/multicluster-observability-addon#44](https://github.com/rhobs/multicluster-observability-addon/pull/44)

SGTM. We already have too many configmaps/secrets to consider per `ClusterLogForwarder` and `OpenTelemetryCollector` stanza.

---
title: multicluster-logs-traces-forwarding
authors:
  - "@periklis"
  - "@JoaoBraveCoding"
  - "@pavolloffay"
  - "@iblancasa"
reviewers:
  - "@alanconway"
  - "@jcantrill"
  - "@xperimental"
  - "@jparker-rh"
  - "@berenss"
  - "@bjoydeep"
approvers:
  - "@alanconway"
  - "@jcantrill"
  - "@xperimental"
api-approvers:
  - "@alanconway"
  - "@jcantrill"
  - "@xperimental"
creation-date: 2023-11-28
last-updated: 2023-11-28
tracking-link:
  - https://issues.redhat.com/browse/OBSDA-356
  - https://issues.redhat.com/browse/OBSDA-393
  - https://issues.redhat.com/browse/LOG-4539
  - https://issues.redhat.com/browse/OBSDA-489
see-also:
  - None
replaces:
  - None
superseded-by:
  - None
---

> # Multi-Cluster Logs and Traces Forwarding

Reviewer 2: The tracing team has concerns with moving forward with this proposal.

Our requirement is to provide OpenTelemetry features in multi-cluster environments which was facilitated by PR - https://github.com/rhobs/multicluster-observability-addon/pull/39#issuecomment-2065235658 and caused confusion about the support responsibilities and was halted for a long time without any proposal how to move it forward.

The observability organization is split into separate teams that develop, maintain, and support a set of operators and CRDs that provide in-cluster user-facing APIs. This proposal enables the distribution and reconciliation of individual CRDs (e.g. `ClusterLogForwarder`) to spoke clusters. Users have to understand and use these CRDs in both in-cluster and multi-cluster setups.

The proposal is to change `mcoa.openshift.io/signal={logging,tracing}` to `mcoa.openshift.io/kind={ClusterLogForwarder,OpenTelemetryCollector}` and structure the addon around kinds/CRDs rather than signals. This makes it straightforward which CRDs users interact with, support questions, and extensibility for the future.

In the arch calls we talked as well about integrating cluster observability operator (COO) into the addon for providing store capabilities in the spokes, which is a cross-signal operator, hence it fits nicely with the kind/CRD structure.

Author: Solving this as since this comment was made progress has been made in https://docs.google.com/document/d/1Yha0IvwmTOzc3HMqG0gj9edRjlPIzIAwvotIFqCBJ9E to ensure MCOA fits all the use cases

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

Multi-Cluster Observability has been an integrated concept in Red Hat Advanced Cluster Management (RHACM) since its inception but only incorporates one of the core signals, namely metrics, to manage fleets of OpenShift Container Platform (OCP) based clusters (See [RHACM Multi-Cluster-Observability-Operator (MCO)](rhacm-multi-cluster-observability)). The underlying architecture of RHACM observability consists of a set of observability components to collect a dedicated set of OCP metrics, visualizing them and alerting on fleet-relevant events.

> This enhancement proposal seeks to bring a unified approach to collect and forward logs and traces from a fleet of OCP clusters based on the RHACM addon facility (See Open Cluster Management (OCM) [addon framework](ocm-addon-framework)) by enabling these signals events to land on third-party managed storage solutions (e.g. AWS Cloudwatch, Google Cloud Logging, Splunk). The proposed addon is an optional RHACM component. It is installed by the MCO Operator. It provides a unified installation approach of required dependencies (e.g. operator subscriptions) and resources (custom resources, certificates, CA Bundles, configuration) on the managed clusters to collect and forward logs and traces. The addon's name is MultiCluster-Observability-Addon (short MCOA).

Reviewer 3: ```suggestion
This enhancement proposal seeks to bring a unified approach to collect and forward logs and traces from a fleet of OCP clusters based on the RHACM addon facility (See Open Cluster Management (OCM) [addon framework](ocm-addon-framework)) by enabling these signals to land on third-party managed storage solutions (e.g. AWS Cloudwatch, Google Cloud Logging, Splunk). The proposed addon is an optional RHACM component. It is installed by the MCO Operator. It provides a unified installation approach of required dependencies (e.g. operator subscriptions) and resources (custom resources, certificates, CA Bundles, configuration) on the managed clusters to collect and forward logs and traces. The addon's name is MultiCluster-Observability-Addon (short MCOA).
```

## Motivation

The main goal here is to simplify and unify the installation of logs and traces collection and forwarding on a RHACM managed fleet of OCP clusters. The core function of the addon is to install required operators (i.e. [Red Hat OpenShift Logging](ocp-cluster-logging-operator) and [Red Hat OpenShift distributed tracing platform](opentelemetry-operator)), configure required custom
resources (i.e. `ClusterLogForwarder`, `OpenTelemetryCollector`) and reconcile per-cluster-related authentication resources (i.e. `Secrets`, `ConfigMaps`). This enables fleet-wide control of logs and traces collection and forwarding.

> ### User Stories

Reviewer 4: Thanks @iblancasa, this set of user stories do seem to me a bit outdated from the latest discussions we had, also taking into account that this initiative started from logging, so it's completely understandable that, from that point of view, the same _could_ be expected from tracing. Apart from that, things have evolved a lot since this initiative started, and our support in the Red Hat build of OpenTelemetry has expanded. Hence the user stories could benefit of redefinition without really impacting the mechanisms to make it happen. I'm also adding the therd part of the regular user story definition: "SO THAT" to help understanding the bigger picture and what we are trying to achieve here.

My 2 cents on what I would expect from a product perspective:
* AS A fleet administrator, I WANT TO install, configure and manage an OpenTelemetry collector on any set of RHACM managed OCP clusters SO THAT I can collect, process and export the signals already available in the Red Hat build of OpenTelemetry in a homogeneous way.

Something very similar can easily apply to the authentication of the OpenTelemetry collector.

I hope this helps! Thanks!

> ### User Stories

Author: I feel the current user stories are abstract enough to encapsulate the user story mentioned here, hence I'm resolving this conversation, if you disagree and want me to explicitly add this user story feel free to unresolve this conversation.

2. The fleet administrator creates a default `ClusterLogForwarder` stanza in the `open-cluster-management` namespace that describes the list of log forwarding outputs. This stanza will then be used as a template by MCOA when generating the `ClusterLogForwarder` instance per managed cluster.
3. The fleet administrator creates a default `OpenTelemetryCollector` stanza in the `open-cluster-management` namespace that describes the list of trace receivers, processors, connectors and exporters. This stanza will then be used as a template by MCOA when generating the `OpenTelemetryCollector` instance per managed cluster.
4. For each managed cluster the MCOA or the fleet administrator provides on the managed cluster namespace additional configuration resources:
   1. For both Logs and Traces a `ConfigMap` with the authentication methods that should be used for each Log Output/ Trace Exporter. 
   2. Per Log Output / Trace Exporter `ConfigMap`: For each output a resource holding output specific configuration (See [ClusterLogForwarder Type: OutputTypeSpec](ocp-clusterlogforward-outputtypespec), [OpenTelemetry Collector Authentication](opentelemetry-collector-auth))

>    3. Each resource created should be annotated with a special label `mcoa.openshift.io/signal` that should be set to the resource that should use that resource.

Reviewer 3: This sentence is unclear to me.  `should be set to the resource that should use that resource` is confusing and I don't really understand what purpose the label should serve.

Author: This was part of the old design this was removed in the new design

   - group: logging.openshift.io
     resource: clusterlogforwarders
     # The default config is the main stanza of a ClusterLogForwarder resource
     # that describes where logs should be forwarded for all managed cluster.
     defaultConfig:
       name: instance
       namespace: open-cluster-management

   # Describes the default OpenTelemetryCollector type applied to all managed clusters.
   - group: opentelemetry.io
     resource: opentelemetrycollectors
     # The default config is the main stanza of an OpenTelemetryCollector resource
     # that describes where traces should be forwarded for all managed cluster.
     defaultConfig:
       name: spoke-otelcol
       namespace: open-cluster-management
```

and a default `AddonDeploymentConfig` describing the operator subscriptions to be used on each managed cluster:

```yaml
apiVersion: addon.open-cluster-management.io/v1alpha1
kind: AddOnDeploymentConfig
metadata:
  name: multicluster-observability-addon
  namespace: open-cluster-management
spec:
  customizedVariables:
    - name: clusterLoggingChannel
      value: stable-5.9
    - name: otelChannel

>       value: stable

Reviewer 5: This is not needed. We only support one channel.

        name: instance
        namespace: open-cluster-management
      spec:
        outputs:
        - loki:
            labelKeys:
            - log_type
            - kubernetes.namespace_name
            - kubernetes.pod_name
            - openshift.cluster_id
          name: application-logs
          type: loki
          url: "https://mycompany.grafana.com/api/logs/v1/spoke" # Pulled from the `logging-spoke-application-logs` configmap
          secret:
            name: spoke-application-logs
        - name: cluster-logs
          type: cloudwatch
          cloudwatch:
            groupBy: logType
            region: us-east-1 # Pulled from the `logging-spoke-cluster-logs` configmap
            groupPrefix: a-prefix # Pulled from the `logging-spoke-cluster-logs` configmap
          secret:
            name: spoke-cluster-logs
        pipelines:
        - inputRefs:
          - application
          name: application-logs
          outputRefs:
          - application-logs
        - inputRefs:
          - audit
          - infrastructure
          name: cluster-logs
          outputRefs:
          - cluster-logs
```

#### Multi Cluster OTLP Collection and Forwarding

> For all managed clusters the fleet administrator is required to provide a single `OpenTelemetryCollector` resource stanza that describes the OpenTelemetry forwarding configuration for the entire fleet in the default namespace `open-cluster-management`.

Reviewer 5: ```suggestion
For all managed clusters the fleet administrator is required to provide, at least, a single `OpenTelemetryCollector` resource stanza that describes the OpenTelemetry forwarding configuration for the entire fleet in the default namespace `open-cluster-management`.
```

We want to support the creation of multiple `OpenTelemetryCollector` instances in near future.

- MCOA manifest sync: with MCOA being deployed by MCO we will need to set up a procedure to maintain the MCOA manifests that live in the MCO repo up to date.
- CRD conflicts: MCOA will leverage the CRDs from other operators we will have to ensure that we will not be running into situations where two operators are managing the same CRD

## Design Details

### Open Questions [optional]

The [addon-framework](ocm-addon-framework) is still lacking some features that will fully enable the behaviours described in this proposal namelly: 
- [ ] Support for reconcile resources generated by the addon (Jira issues [LOG-5343](https://issues.redhat.com/browse/LOG-5343) and [TRACING-4148](https://issues.redhat.com/browse/TRACING-4148)). MCOA needs this otherwise it will not react on changes to those resources.
- [ ] Support for multiple resources of the same kind in the ManagedClusterAddOn resource (Jira issues [LOG-5344](https://issues.redhat.com/browse/LOG-5344)). MCOA needs this otherwise it will not react on changes to those resources.

### Test Plan

TBD

### Graduation Criteria

TBD

#### Dev Preview

- Installation of the ClusterLogForwarder and OpenTelemetryCollector CRs.

> - Support for all the authentication methods supported by ClusterLogForwarder (i.e. static auth, mTLS auth, etc).

Reviewer 5: ```suggestion
- Support for all the authentication methods supported by ClusterLogForwarder and Red Hat Build of OpenTelemetry (i.e. static auth, mTLS auth, etc).
```

> - Support for Per-ClusterSet ClusterLogForwarder.

Reviewer 5: ```suggestion
- Support for Per-ClusterSet ClusterLogForwarder and OpenTelemetryCollector.
```

- Health probing per Signal
- Status reporting

#### Dev Preview -> Tech Preview

- Refined CLF/OTelCollector CRs usage

#### Tech Preview -> GA

TBD

#### Removing a deprecated feature

None

### Upgrade / Downgrade Strategy

None

### Version Skew Strategy

None

### Operational Aspects of API Extensions

TBD

#### Failure Modes

> - Cert Manager not installed: this will not allow the addon to create it's own CA and provision mTLS certificates

Reviewer 5: Maybe it is an antipattern but, since we have access from the addon to the Hub cluster... would it be crazy to install the `cert-manager` from the addon if not present?

| MCOA |  ✅  |   ✅  |      🚧     |             |                      |

POC - https://github.com/ViaQ/logging-ocm-addon
Demo - https://github.com/rhobs/multicluster-observability-addon/tree/main/demo
DevPreview - https://github.com/rhobs/multicluster-observability-addon

## Alternatives

### Policy-Engine based workflow

The RHACM policy engine is part of the RHACM governance tools to apply certain policies on workloads and infrastructure on the entire fleet. It serves well to a degree however for the MCOA use-case it falls short since MCOA will need to:
- manage complext dependencies between multiple components, OpenShift Logging Operator installation (i.e Cluster-Logging-Operator) and the custom resources (i.e. ClusterLogForwarder) as well as authentication related artifacts (i.e. generating TLS certificates, Cloud Provider Managed Identities / ServiceAccounts, etc.).
- build dynamic error reporting (e.g. operator installation/upgrade failures, CRD handling) and health reporting (e.g. running/ready progress of operators and operands).

The spike for this alternative can be found here: https://gitlab.cee.redhat.com/openshift-logging/log-storage-toolbox/-/merge_requests/14

> ### Integrate directly into present Multi-Cluster-Observability-Operator

Reviewer 5: This is the final decision, right?

Author: This alternative relates to the idea of instead creating MCOA just updating the MCO code directly.

## Infrastructure Needed [optional]

None

[ocm-addon-framework]:https://github.com/open-cluster-management-io/addon-framework
[ocp-cluster-logging-operator]:https://github.com/openshift/cluster-logging-operator
[ocp-clusterlogforwarder-outputsecretspec]:https://github.com/openshift/cluster-logging-operator/blob/627b0c7f8c993f89250756d9601d1a632b024c94/apis/logging/v1/cluster_log_forwarder_types.go#L226-L265
[ocp-clusterlogforward-outputtypespec]:https://github.com/openshift/cluster-logging-operator/blob/627b0c7f8c993f89250756d9601d1a632b024c94/apis/logging/v1/output_types.go#L21-L40
[opentelemetry-collector-auth]:https://opentelemetry.io/docs/collector/configuration/#authentication
[opentelemetry-operator]:https://github.com/open-telemetry/opentelemetry-operator
[rhacm-multi-cluster-observability]:https://github.com/stolostron/multicluster-observability-operator

> ## RANDOM  IDEAS 

Reviewer 5: ```suggestion
## RANDOM  IDEAS 
- Support the creation of multiple OpenTelemetry Collector instances instead of just one to export the telemetry data outside the cluster
- Support TempoStack to store some of the traces locally in the cluster
```

Reviewer 2: Instrumentation support, meta-monitoring of collectors

- What happens if we need to reference multiple static secrets?

> - Maybe we can use annotations for the authentication methods instead of having to create a seperate resouce for that

Reviewer 5: What advantages do we currently have with the separate resource? I think this can be benefitial for users if it simplifies the workflow.
